### PR TITLE
Customize Twitter/Facebook cards and SEO meta data.

### PR DESCRIPTION
**Pull request by conroyp**
This PR adds the ability to customise the meta description for a post, as well as the meta data used to format twitter/facebook cards when sharing on either network. It allows for an author to target users of a particular social network with an optimal title/description/image, providing backend support for an [Open Graph](http://ogp.me/) and [Twitter card](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/guides/getting-started.html) implementation on the front end.

### UI changes
Social meta edit modal

I've added settings to allow overriding of:
* meta description
* twitter title, description and image
* facebook (opengraph) title, description and image

Adding these directly to the existing settings modal makes it very long and unwieldy, so instead I added a tabset to the settings dialog. This allows a grouped separation between the existing settings, and these new seo/social ones. I took this approach rather than adding a new settings modal, as my thinking is that these options belong close to the slug/excerpt editing. Within the social tab, the Facebook and Twitter fields are contained within an accordion, again with the idea of preventing the user from getting bombarded with too many inputs at once on initial load!

The logic for both the twitter and facebook field editing is pretty much identical, so a new `SocialMeta` vue component has been added to prevent duplication within the main `edit.vue` file.

One thing to note here is that label shown to users on the Facebook accordion panel is "Facebook", but the input fields are named `opengraph_*`. The thinking here is that users will be looking for the Facebook fields to edit, but from a dev perspective, meta field names need to be outputted as `opengraph_*` on the frontend, not `facebook_*`, so under the hood the field names are kept consistent. This is a similar approach to the one taken by Yoast on the Wordpress SEO plugin.

### Backend changes

As part of this PR, a `WinkMeta` model has been added. This is similar to wordpress' postmeta, allowing us to store key, value and link to a post. The main benefit of this is that it makes it a lot easier to add new meta fields in future, without having to make schema changes. The downside of this is that by default, the api response that comes back when `with('meta')` is added comes back something like:
```
meta: [
    [id: 1, key: 'opengraph_title', value: 'My FB title', created_at....],
    [id: 2, key: 'opengraph_description', value: 'My FB description', created_at....],
    ....
]
```
This isn't the easiest structure to work with on the front end. There we'll usually want to know things like "is `opengraph_title` set on this post?", and having to iterate over a collection each time when looking to get/set isn't ideal. To get around this, json resources are added for both the post and post meta models, with collections for each also present. The resource formatting doesn't do anything special with the post, other than to wrap up the attached meta fields, so the existing structure of the post response shouldn't change outside the addition of a `meta` field. This allows a meta response like the above to be converted into something hopefully a bit more developer-friendly, like the below:

Api fields

One other thing that may look a bit funny here is the way that `opengraph_image_width` and `opengraph_image_height` are calculated whenever `opengraph_image` is updated. The goal here is to avoid a known issue with the facebook share crawler whereby the first user to share a piece of content often doesn't see the image in their preview. Explicitly specifying the height and width is the recommended solution for this (see https://developers.facebook.com/docs/sharing/best-practices#precaching). Rather than ask users for these values, they can be extracted from the image on upload, keeping them in the db for whenever they're needed for frontend rendering.

### CODE REVIEW

**Comment by Lloople** [comment]
What a huge PR! 👏 I hope it's get merged, looks pretty dope 🙂

**Comment by themsaid** [comment]
@conroyp It looks so good. However, I was hoping to make the move to Tailwind before people starting to make such PRs or help me with the road map. You were faster than myself :D

I just merged a huge change that moves us from Bootstrap to TailwindCSS. I find myself overriding lots of bootstrap classes so thought it would be better to start from a clean CSS state which Tailwind provides.

I did a lot of changes that made the styling more clear and easier to change in the future. But that means merging this PR wouldn't be easy.

If it's possible can you try and see if you can merge with master? If not please let me know and I'll try to find sometime to merge the PR.

THANKS A TON ❤️

**Comment by themsaid** [comment]
Instead of Using tabs I'd put all the settings in the same place, later one we can have the modal styles in a better way so it has multiple screens similar to swiping on the phone. Thing with tabs is that you're limited by the width of the modal, you can add more tabs or you'll break the layout.

So just before the "Delete This Post" button, you may add the new section for facebook & twitter.

Regarding the structure, how about using a JSON `meta` field instead of adding a meta table? That way we can store the JSON structure in a key-value fashion. That's the approach I took with Telescope and it turned out pretty decent.

**Comment by conroyp** [comment]
@themsaid thanks for the quick feedback and suggestions - I'll take a look at getting the conflicts with tailwind sorted out, and moving to a JSON `meta` field, rather than a new table.

**Comment by conroyp** [comment]
@themsaid I've caught up with master, and fixed the conflicts whilst killing off the tabs. I've left the social sharing settings in expandable cells which default to closed - I'd expect the common case will be users leaving the defaults as-is unless there's a special case to change, in which case it's a lot of extra fields to be showing on every load of the settings screen. Let me know if I'm off-base on that assumption, and you'd rather them all be shown all the time.

I've also ripped out the meta table, changing it to the json field you mentioned. Casting this on the post model means that all of the resource stuff becomes pretty redundant, so that's been pulled back out now also.

File: src/WinkPost.php [hunk 1]
```diff
@@ -50,6 +50,15 @@ class WinkPost extends Model
         'publish_date'
     ];
 
+    /**
+     * The attributes that should be casted.
+     *
+     * @var array
+     */
+    protected $casts = [
+        'meta' => 'array'
```
Comment (Lloople) [inline src/WinkPost.php]: Just asking.

Why is this needed? The column is already json 🤔

Great work in this PR man 🙌

Comment (conroyp) [reply src/WinkPost.php]: @Lloople thanks very much, appreciate it, and the quick attention to this PR 😃 

This cast here is intended mainly for convenience, both on backend and front. Without it, the api response for the meta field looks not like it does in the screenshots above, but more like the below:
image
The front end would then need to json decode the meta data before assigning to properties in the views.

Similarly, on the front end each of the fields is set up as `form.meta.NAME`, and posted through the standard api post update code. Without the json field casting on the post model in the backend, this flow throws an error when the store is done, as we're trying an array to single conversion.
The front end could be updated to stringify the `form.meta` data before submission to get around this also (or do something similar in the controller), but by adding the casts value, we get to treat `meta` like an array on the way in and out, without having to write any conversion functions ourselves anywhere else in the code.

**Comment by themsaid** [comment]
Thanks :) will review and merge the soonest :)

**Comment by conroyp** [comment]
@themsaid I saw a merge conflict pop up again this morning - went to resolve it and the accordion appears to have stopped working. You can hold off reviewing it for the time being - I'll ping you once I get that ironed out!

**Comment by conroyp** [comment]
@themsaid conflicts resolved now, this one should be good for another look whenever you get a chance.

**Comment by themsaid** [comment]
@conroyp thank you so much for your work. I've updated the styles and gave us a chance to add more things later.

Can you please pull from master and try it out? The opengraph image thing has be omitted just for now but I plan to add it when we apply the SEO/Social meta thing to the rest of entities.

So I was thinking of applying the same to Authors, Tags, and Pages. So you can control the look of the cards when their screens are shared on social media. What do you think?

**Comment by conroyp** [comment]
@themsaid that design change looks great - I really like the split now within the settings icon. The indicator on the image upload is a really nice touch too 👍 

Adding to the tags, authors and pages sounds like a good idea too. Would you like me to take that on?
On pages, I imagine the settings icon would generate a menu more in line with posts. On authors/tags, they're essentially just large forms at the moment. Would you see the additional fields dropping in here as new rows on those forms, or the beginning of a breakout into tabs/additional settings dialog?

**Comment by themsaid** [comment]
@conroyp would be great if you can help me with that :)

In Authors and Tags, I'd add a settings icon same as posts/pages and show a dropdown with only the "SEO & Social" item.
